services/QuillLMS/app/services/classlink_integration/user_persister.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module ClasslinkIntegration
+  class UserPersister < ApplicationService
+    attr_reader :incoming_user, :existing_provider_users, :classlink_instance_id
+
+    def initialize(incoming_user:, existing_provider_users:, classlink_instance_id:)
+      @incoming_user = incoming_user
+      @existing_provider_users = existing_provider_users
+      @classlink_instance_id = classlink_instance_id
+    end
+
+    def run
+      creatable_user_hash = UserPersister.creatable_user(incoming_user)
+
+      if existing_user
+        existing_user.user.update!(**creatable_user_hash)
+      else
+        ActiveRecord::Base.transaction do
+          User.transaction do
+            user = User.create!(**creatable_user_hash)
```

Comment:
I think there may be an edge case here where we'll get validation failures if a non-classlink user account already exists in the Quill database for the email address we get from Classlink.  In that case, `existing_user` won't find a result (since there's no matching `external_id`), so `create!` will try to make a new user with a duplicate email.